### PR TITLE
Enhance the otelcol helm chart to support windows nodes

### PR DESCRIPTION
**Pull request by nitaliya**
Description:
Helm chart changes to make support of k8s cluster with windows worker node.
Ticket : [OTL-535](https://signalfuse.atlassian.net/browse/OTL-535)
Following activities/changes done in this PR.

- Added flag `isWindows` for windows/linux toggle.
-  `kubeletstats` receivers in default configuration. because function x509.SystemCertPool() used here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/kubelet/cert.go#L25) is not available for windows. Reference - https://github.com/golang/go/issues/16736
So we have made changes in `opentelemetry-collector-contrib` repo for creating an empty CertPool when the system run environment is windows. [PR Link](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5255)
- Added `nodeSelector:` `kubernetes.io/os: windows` to schedule pod on a windows worker node only.
- Change the mount path of volumes for windows mode.
- Added `initialDelaySeconds` for `readinessProbe` and `livenessProbe` because Windows pods are taking some time for spinning up.

### CODE REVIEW

**Comment by dmitryax** [comment]
Without container metrics, it'll be pretty limited functionality.  Does signalfx-agent collect container metrics?

**Comment by nitaliya** [comment]
> Without container metrics, it'll be pretty limited functionality. Does signalfx-agent collect container metrics? If so, why doesn't it face the same issue with crypto/x509?

In signalfx-agent we are creating a new CertPool when the system run environment is windows - [reference](https://github.com/signalfx/signalfx-agent/blob/main/pkg/core/common/auth/tls.go#L62)

**Comment by dmitryax** [comment]
Can we apply the same approach CertPool in contrib or splunk-otel-collector?

I don't think we want to provide k8s agent that cannot collect container metrics. That is essential data.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Fixed
 
 - smartagent/kubernetes-events now works with `otelK8sClusterReceiver` deployment. (#187)
+- Support adde for k8s cluster with windows worker node. (#190)
```
Comment (dmitryax) [inline CHANGELOG.md:10]: This should be moved to ### Added section.
```suggestion
- Support k8s clusters with Windows nodes (#190)
```

File: examples/README.md [hunk 2]
```diff
@@ -17,6 +17,26 @@ splunkRealm: us0
 splunkAccessToken: my-access-token
 ```
 
+## Deploy for k8s cluster with windows worker node.
+
+This configuration will install application for windows worker node only.
+
+All windows images are available in a separate `quay.io` repository for that we need to change the image pul repository to `quay.io/signalfx/splunk-otel-collector-windows`.
```
Comment (dmitryax) [inline examples/README.md:24]:
```suggestion
All windows images are available in a separate `quay.io` repository: `quay.io/signalfx/splunk-otel-collector-windows`.
```

File: examples/README.md [hunk 3]
```diff
@@ -17,6 +17,26 @@ splunkRealm: us0
...
+
+Windows node only supports traces and metrics collection. Because As of now, we don’t have fluentd-hec image for the windows worker node,  as we are moving away from fluentd, No logs will be collected.
```
Comment (dmitryax) [inline examples/README.md:26]:
```suggestion
Windows support covers traces and metrics collection only, no logs will be collected. As of now, we don’t have fluentd-hec image for the windows worker node.
```

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml [hunk 4]
```diff
@@ -39,13 +39,23 @@ spec:
         sidecar.istio.io/inject: "false"
         {{- end }}
     spec:
+      {{- if not .Values.isWindows }}
       hostNetwork: true
+      {{- end }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/daemonset.yaml:44]: This will affect host network metrics. What prevents us from keeping it? Can signalfx-agent run in hostNetwork on Windows nodes?

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml [hunk 5]
```diff
@@ -39,13 +39,23 @@ spec:
...
       dnsPolicy: ClusterFirstWithHostNet
       serviceAccountName: {{ template "splunk-otel-collector.serviceAccountName" . }}
+      {{- if .Values.isWindows }}
+      nodeSelector:
+        kubernetes.io/os: windows
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/daemonset.yaml:49]: Since we already has bunch of custom configuration requirements for Windows in README.md. Let's move this part there as well.

File: helm-charts/splunk-otel-collector/values.yaml [hunk 6]
```diff
@@ -674,6 +681,9 @@ otelCollector:
   # existing fields can be disabled by setting them to `null`.
   config: {}
 
+# Specifies whether to apply for k8s cluster with windows worker node.
+isWindows: false
+
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/values.yaml:686]: Let's move this part under `Extra system configuration` section

**Comment by dmitryax** [comment]
Also please rebase

File: helm-charts/splunk-otel-collector/values.yaml (hunk 6 above)
Comment (nitaliya) [reply helm-charts/splunk-otel-collector/values.yaml]: Moved `isWindows` flag.

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml (hunk 5 above)
Comment (nitaliya) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml]: For `nodeSelector` i have taken [reference](https://github.com/signalfx/signalfx-agent/blob/main/deployments/k8s/helm/signalfx-agent/templates/daemonset.yaml#L60) of smart-agent. in which value is setting based on the `isWindows` flag.
If you suggest then we can remove this part from confirmation and put it in README.md

Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml]: new version looks better

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml (hunk 4 above)
Comment (nitaliya) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml:44]: We can not set `hostNetwork: true` for the windows worker node because `/etc/cni/net.d` and its `/opt/cni/bin` will note be present on the host. so we are getting the following error for that.

Screenshot 2021-09-20 at 10 24 42 AM

Also, for SignalFx-agent we are not keeping it in case of windows installation - [Reference](https://github.com/signalfx/signalfx-agent/blob/main/deployments/k8s/helm/signalfx-agent/templates/daemonset.yaml#L38)

**Comment by nitaliya** [comment]
Hi @dmitryax 
As you suggested I have added `metricstransform` processor to rename `container.memory.working_set` metric to `container_memory_usage_bytes` and also validated `Kubernetes navigator dashboard`, now it is collecting all matrices.
Can review this? and let me know if any changes are required.

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml (hunk 4 above)
Comment (dmitryax) [reply helm-charts/splunk-otel-collector/templates/daemonset.yaml:44]: ok. sounds good

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl [hunk 7]
```diff
@@ -198,6 +198,14 @@ processors:
         value: "{{ .Values.environment }}"
   {{- end }}
 
+  {{- if .Values.isWindows }}
+  metricstransform:
+    transforms:
+      - include: container.memory.working_set
+        action: insert
+        new_name: container_memory_usage_bytes
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl:206]: I believe this metrics represented as `container.memory.usage` in otel conventions, and will be translated on the backend to `container_memory_usage_bytes` if needed. please validate usage of `container.memory.usage` instead.

File: helm-charts/splunk-otel-collector/templates/daemonset.yaml [hunk 8]
```diff
@@ -39,13 +39,17 @@ spec:
...
       dnsPolicy: ClusterFirstWithHostNet
       serviceAccountName: {{ template "splunk-otel-collector.serviceAccountName" . }}
-      {{- with .Values.nodeSelector }}
       nodeSelector:
-        {{ toYaml . | nindent 8 }}
-      {{- end }}
+        {{- if .Values.nodeSelector }}
+        {{ toYaml .Values.nodeSelector | nindent 8 }}
+        {{- else }}
+          kubernetes.io/os: {{ .Values.isWindows | ternary "windows" "linux" }}
```
Comment (dmitryax) [inline helm-charts/splunk-otel-collector/templates/daemonset.yaml:51]: just a nit
```suggestion
        kubernetes.io/os: {{ .Values.isWindows | ternary "windows" "linux" }}
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -16,6 +16,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- smartagent/kubernetes-events now works with `otelK8sClusterReceiver` deployment. (#187)
```
Comment (dmitryax) [inline CHANGELOG.md:19]: This is unrelated change

Comment (dmitryax) [reply CHANGELOG.md:19]: Please rebase on the latest main

File: helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl (hunk 7 above)
Comment (nitaliya) [reply helm-charts/splunk-otel-collector/templates/config/_otel-agent.tpl]: It also works with `container.memory.usage`, so I have made this change.

File: CHANGELOG.md (hunk 9 above)
Comment (nitaliya) [reply CHANGELOG.md]: Done. CHANGELOG.md is now updated.

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
Should be merged after this change is available in the splunk-otel-collector docker image
